src/simd/definitions.h

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2017-2018 Scality
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/** @file definitions.h
+ *
+ * Provide includes and constants that are platform-specific.
+ */
+
+#ifndef __QUAD_SIMD_SIMD_DEFINITIONS_H__
+#define __QUAD_SIMD_SIMD_DEFINITIONS_H__
+
+#include <climits>
+#include <cstdint>
+
+// SIMD-specific includes {{{
+
+#ifdef QUADIRON_USE_SIMD
+
+// SIMD for x86 and x86_64
+#if defined(__i386__) || defined(__x86_64__)
+#include <x86intrin.h>
+#endif
+
+// SIMD for ARM
+#if defined(__ARM_NEON)
+#include <arm_neon.h>
+#endif
+
+#endif
+
+// }}}
+
+namespace quadiron {
+namespace simd {
+
+/// Supported instruction set.
+enum class InstructionSet {
+    NONE, ///< No SIMD instruction (fallback).
+    SSE,  ///< SSE4.1
+    AVX,  ///< AVX2
+};
+
+// Definitions for Intel AVX-256 {{{
+
+// We required AVX2 because we relies on some instructions (such as
+// `_mm256_add_epi16` and others) that aren't available in the first version of
+// AVX.
+#if defined(__AVX__) && defined(__AVX2__)
+
+using RegisterType = __m256;
```

Comment:
We have `MaskType`, so do we need this type?

Reply:
`MaskType` has a different type than `RegisterType`, and in my future PR we will have two different classes (`Register` and `Mask`).
A mask and a register have different interface and constraint, like a mask must be an integer, a register can hold floating-point value (not in our case here in QuadIron, but still I think it's cleaner to separate them: anyway the conversion/cast between those two types are free (no instruction are generated)).

Reply:
It's still not clear for me :)
I think we work only for integers, not floating-point variables

Reply:
Maybe it will be clearer in the next PRs, when I'll actuelly use it :)

>  I think we work only for integers, not floating-point variables

Yes, in our very specific case we only use integers (for now), but it doesn't hurt to have distinct types for distinct uses (multiplying two mask together doesn't make sense, so you can implement `operator *` for RegisterType and not for `MaskType`, etc).

Reply:
yes, I'd like to see its use in next PRs